dynawo/sources/Solvers/AlgebraicSolvers/DYNSolverKINAlgRestoration.cpp

```diff
@@ -54,65 +61,52 @@ mode_(KIN_NORMAL) {
 }
 
 SolverKINAlgRestoration::~SolverKINAlgRestoration() {
-  clean();
+  cleanAlgebraicVectors();
 }
 
 void
-SolverKINAlgRestoration::init(const shared_ptr<Model>& model, modeKin_t mode, double fnormtol, double initialaddtol, double scsteptol,
-                              double mxnewtstep, int msbset, int mxiter, int printfl) {
-  // (1) Arguments
-  // --------------
+SolverKINAlgRestoration::cleanAlgebraicVectors() {
+  if (sundialsVectorY_ != NULL) {
+    N_VDestroy_Serial(sundialsVectorY_);
+    sundialsVectorY_ = NULL;
+  }
+}
+
+void SolverKINAlgRestoration::resetAlgebraicRestoration() {
   clean();
+  cleanAlgebraicVectors();
+  numF_ = 0;
+}
 
+void
+SolverKINAlgRestoration::init(const shared_ptr<Model>& model, modeKin_t mode) {
   model_ = model;
   mode_ = mode;
-  initVarAndEqTypes();
-
-  if (nbF_ == 0)
-    return;
-
-  switch (mode) {
-    case KIN_NORMAL:
-      initCommon("KLU", fnormtol, initialaddtol, scsteptol, mxnewtstep, msbset, mxiter, printfl, evalF_KIN, evalJ_KIN);
-      break;
-    case KIN_YPRIM:
-      initCommon("KLU", fnormtol, initialaddtol, scsteptol, mxnewtstep, msbset, mxiter, printfl, evalF_KIN, evalJPrim_KIN);
-      break;
-  }
 }
 
-
-void
+bool
 SolverKINAlgRestoration::initVarAndEqTypes() {
```

Comment:
What about updating the method name as it does both updating the types of the variables and equations, but also initializing ignore and index vectors and more importantly, indicates if we need to rebuild the KINSOL structure. 
I'm not very sure that the method should return a boolean itself or just return a new numF. In the latter case, the comparison could directly be done in the setUp method...

Reply:
I propose a new implementation in my next commit tell me if you prefer